### DISCUSSION TITLE
Page templates - Beautiful by default, automatic content discovery, menu building and search service

### DISCUSSION THREAD

**Discussion by Alyxion** [zauberzeug/nicegui#2846] [Ideas / Feature Requests]
Inspired by the discussion about page styling and site content provision in [#2811](https://github.com/zauberzeug/nicegui/pull/2811) and thinking about how to make NiceGUI even more convenient and how to even further lower the level of "building a cool app" without any prior web dev knowledge in the long term.

As you proved with https://nicegui.io you **can** build really beautiful apps with NiceGUI, after all NiceGUI grants full control of the HTML, CSS etc. But I would claim the team also invested a lot of time and love into all the details... like the styled fragments, multi colored titles, all these small things which make the difference if a visitor thinks "Wow, thats a beautiful, professional site" and "Hm, ok, works".

For someone building his/her first NiceGUI app there is atm still a quite steep learning curve if it goes beyond just a hello world demo. 

Yes, you can set up a left_drawer, right_drawer etc., yes you can put menus in there for page navigation, yes you **could** build your own search functionality etc. but I think we agree for someone completely new to NiceGUI and web development this will take days if not weeks.

Thinking out of the box of the current hard wired @page approach to provide page content, inspired by how Material for MkDocs currently solves this:

* main.py - Just creating some SinglePageApp and PageBuilder object and defining a config (darkmode, viewport, title etc.)
* pages/
    * index.md - My Markdown landing page pointing to my awesome apps, rendered via ui.markdown
    * mail - Mail sub app
        * index.py- Mail inbox - no single decorator required at all, module is executed top to bottom by default, so basically just import nicegui.ui and go for it. The page title is by default derived from the file name.
    * calendar - Calender sub app
        * new_appointment.py
        * index.py - General landing page
        
If the **system** - such as a page builder class - would be able to discover and collect the content from "the whole site" by itself, so

* the page names and hiearchy
* the page sections (still requiring some standard implementation of "id" as component) like you did on NiceGUI.io:
![image](https://github.com/zauberzeug/nicegui/assets/33489959/4148fb34-0c26-4eb1-82eb-5293e222de32)
* the searchable content - e.g. by doing an invisible build of each page or providing a custom search handler for more experienced users

... then everybody could build a beautiful app w/ global site navigation, with content search and in-page navigation with like fifty lines of code.

**Comment by vickorian** [comment]
So this seems to go to #2822 which is what i proposed. There are lots of ways to combine the elements and generate appearances, but there isn't a "bible" per se for how to do it well or efficiently. Each of the pages you proposed would need to be styled differently to fit the themes that people would want to apply to their site. People would want different configurations, different options, etc.

So while i think that pre-rendered pages would be great for new users, like plug and play sites, there would need to be standardization to apply certain effects. And there are so many options, there probably needs to be community effort to make this happen in a timely manner. At least from my perspective.

**Reply by vickorian** [reply]
@Alyxion unless you're specifically referring to the definition of pages as a concept and not the direct styling of said pages.

**Reply by Alyxion** [reply]
Actually **styling details** is the not so important part of this discussion. Yes, as said you said in your second post it is far more about the **pages as a concept**.

* What pages are there?
* In which hierarchical order?
* Do these pages contain sub sections?
* What (searchable) content do they contain?

And when you have all this information plus a lets name it SiteBuilder class (with which the NiceGUI user should not have any direct contact more than like ui.run(site_config='./config.yaml') or something similar, then...

* You can automatically populate a site menu
* Can populate the in-site menu (fragments)
* Can provide a website wide search for the content

This in combination with a "base skeleton" of a professional website once provided by a guy who has styling skill - which is the same for all websites and pages of them, but you can switch features off you don't like and site-builder specific helper classes (such as hyperlinks by default having a professional appearance rather than looking in html 1.0)

Basically its the same quite popular and well working concept of **Material for MkDocs**, just transferred from Markdown files to Python files, having dynamic instead of static content, replacing `# Header` through pb.h1("Header"), replacing ui.label("Subsection").attr("id=section") through with pb.h2("Sub section", "section"), replacing ui.link("...") with something like pb.link("') so that you get a link like this by default: 
![image](https://github.com/zauberzeug/nicegui/assets/33489959/ed2b94c7-55a3-4641-912c-fdd2219d887c) and so on.

I think the current concept of the NiceGUI documentation system isn't even so far away from that concept.

All the user still has to do is provide the content (page by page, without having to know details about page switching, menus etc.), put them in the right folder and be done. Plus a configuration file in the root in which you can define:
* Company logo
* Preferred font if its not the standard one
* Main navigation on top or on the side?
* Primary and secondary colors for bright and dark theme
* Which optional extensions? Mermaid, code highlighting etc.

See here https://squidfunk.github.io/mkdocs-material/setup/

And on a per-page basis you can still do fine tuning then, e.g. hide sidebars if you need the full space, automatically collapse them etc.

**Streamlit** currently has a light version of it. If you create a `pages` folder in your streamlit app and put python files in there, then it will always execute:
* build content of the actual main.py you passed as your entry point to streamlit run
* attach content of what ever pages/{current_page} produced
* attach a menu sidebar which is automatically populated with all menu elements for each page

Their current implementation is really horrible though in my opinion, because as of now:
* Python filename = Site name in the menu, **not** changeable. Thus you end up with Python files names such as `1. 🏠 Home.py` ... in your Git, yes, including the Emoji... awesome.
* No chance to customize or adjust anything per-page at all, so no, you could not even get rid of the sidebar even if you would want.
* Page content defined in the modules context root. I mean I taught several people Python meanwhile and one of the very first things I enforce is "declare at least one function and do not just throw everything on the ground", its just super bad style in my opinion. So yes, at least something like `def content():`. because then you can also just collect the module without harm or automatisms, lookout for something like a config object in that page's module, e.g. defining viewport, real page title if it shall not be the Python filename etc. etc.

**Reply by vickorian** [reply]
@Alyxion So i think that's a great idea, but it might be better to lean more on the typical structure and data types of python. Instead of only providing an external .yaml file which can seem foreign to people and a moderate pain to set up, it might be better to store that hierarchical data in a dictionary that's easily extractable and manipulable from any python environment. 

Or, you just provide the execute at runtime and the code is checked to determine the type of input, either 'str' for a yaml filetype or 'dict' object for runtime defined structure. Since nicegui needs to be run anyways to execute, there's no reason not to provide an option for both. It would also speed up development. If people don't care about the extra options they could quickly get the site up and running with the dictionary method and translate to a yaml as they needed more customization.

But i think it would help a lot if we had some base skeletons for starter sites. I'm working on some things to make AgGrids more accessible and allowing for automatic integrations from certain databases/dataframes. Things that would let you just point them in the right direction and it handles all the importing, manipulation, etc. There are other Intergrations like mail apps, calendar and scheduling, etc that I could see being easily standardized and pre-packaged as pages with certain inherited stylings.

**Comment by rodja** [comment]
I like the general idea a lot. I would also lean towards "all Python" instead of a yaml configuration file. As @Alyxion pointed out the template engine could provide high-level elements like `section`, `subsection`, `text`, `picture`, ... I just wonder if it should be a core part of NiceGUI or some additional package like "nicegui-template".

**Reply by vickorian** [reply]
a separate package seems pretty reasonable. Something that provides a barrier between the base nicegui and the pre-structured content. It would definitely help with requests that users might make, so just from an admin standpoint, it think it's better.

**Reply by Alyxion** [reply]
Agree with @vickorian. Was the same for kivy with it's kivy-md or Material for MkDocs for MkDocs.
This way NiceGUI stays clean and maintainable.

**Comment by rodja** [comment]
I just stumbled upon a base structure which was posted on Discord: https://github.com/bitdoze/nicegui-starter
